qlib/ConnectionProvider.qm

```diff
@@ -366,34 +375,92 @@ string connstr = get_connection(str);
+            @param options connection options
+
+            @par Attributes
+
+            Attributes are special flags for given connection - stored as public
+            attributes of the object instance.
+            For now there are:
+                - \c enabled, default \c True
+                - \c monitor, default \c True
+
+            @par Safe URL
+
+            Public attribute \c safe_url is created from the \c url calling
+            AbstractConnection::getSafeUrl(). Reimplement this method if your
+            connection does not follow standard URL in form:
+            @code
+            schema://user:pass@hostname:port/path
+            @endcode
+
+            @par URL parsing
+
+            The \c url argument is parsed by AbstractConnection::parseUrl().
+            @ref Qore::parse_url() is called in the base implementation.
+            Reimplement \c parseUrl() if your connection uses different URL schema.
+            Like eg. Qore::Datasource connection string.
+         */
+        constructor(string name, string description, string url, hash attributes = {}, hash options = {}) {
+            constructorInit(name, description, url, attributes, options);
+        }
+
+        private:internal constructorInit(string name, string description, string url, hash attributes, hash options) {
+            # metadata
+            self.name = name;
+            self.desc = description;
+
+            # url handling
+            self.url = url;
+            self.urlh = parseUrl(self.url);
+            self.safe_url = getSafeUrl(self.urlh);
+
+            # attributes to flags
+            self.monitor = attributes.monitor ?? True;
+            self.enabled = attributes.enabled ?? True;
+
+            self.orig_opts = options;
+
+            options = getDefaultOptions() + options;
+            if (options) {
                 *hash oh = getOptions();
-                foreach string opt in (n_opts.keyIterator()) {
+                foreach string opt in (options.keyIterator()) {
                     if (!oh{opt})
                         throw "CONNECTION-ERROR", sprintf("connection %y (type %y url %y) has unknown option %y (known options: %y)", name, urlh.protocol, url, opt, oh.keys());
                 }
             }
-            opts = n_opts;
-            enabled = n_enabled ?? True;
+            self.opts = options;
+        }
+
+        private hash parseUrl(string url) {
+            return Qore::parse_url(url);
         }
 
         #! creates a "safe" URL string with password information removed
-        /** @param urlh broken down URL hash (as returned by @ref Qore::parse_url())
+        /** This base/default implementation expects the \c urlh
+            as returned by @ref Qore::parse_url())
 
             @return a "safe" URL string with password information removed
          */
-        private static string getSafeUrl(hash urlh) {
+        private string getSafeUrl(hash urlh) {
```

Comment:
what's the reason to not have this static anymore? It doesn't manipulate with member values...?

Reply:
because of inheritance. The main goal of this patch is to allow to implement customized getSafeUrl() nad parseUrl() - to be able to inherit them. See:

```
%new-style

class A {
    constructor(string url) {
        printf("%N\n", parseUrl(url));
    }
    static string parseUrl(url) {
        return "A = " + url;
    }
}

class B inherits A {
    constructor() : A("url from b") {
    }
    static string parseUrl(string url) {
       return "B = " + url;
    }
}

A a("url for a");
B b();
/*
"A = url for a"
"A = url from b"   --- the A::parseUrl() is always called
*/

class C {
    constructor(string url) {
        printf("%N\n", parseUrl(url));
    }
    string parseUrl(url) {
        return "C = " + url;
    }
}

class D inherits C {
    constructor() : C("url from d") {
    }
    string parseUrl(string url) {
        return "D = " + url;
    }
}

C c("url from c");
D d();
/*
"C = url from c"
"D = url from d"    --- now it's OK, the inherited version is called
*/
```